volgen/cluster-graph.go

```diff
@@ -0,0 +1,137 @@
+package volgen
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/gluster/glusterd2/brick"
+	"github.com/gluster/glusterd2/volume"
+)
+
+const (
+	clusterGraphType = "cluster.graph"
+)
+
+var (
+	ErrClusterNoChild              = errors.New("cluster nodes cannot have children")
+	ErrInvalidClusterGraphTemplate = errors.New("invalid cluster graph template")
+	ErrIncorrectBricks             = errors.New("incorrect number of bricks given for volume")
+)
+
+func isClusterGraph(voltype string) bool {
+	return voltype == clusterGraphType
+}
+
+func newClusterGraph(a qArgs) (*Node, error) {
+	if len(a.t.Children) != 0 {
+		return nil, ErrClusterNoChild
+	}
+
+	g, err := getTemplate(strings.ToLower(a.vol.Type.String()) + ".graph")
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := processClusterGraph(g.root, a.vol)
+	if err != nil {
+		return nil, err
+	}
+
+	// At the end, we should have a single node in the returned list which we return
+	// If not, we had an incorrect number of bricks for the volume
+	if len(n) != 1 {
+		return nil, ErrIncorrectBricks
+	}
+
+	return n[0], nil
+}
+
+func processClusterGraph(t *Node, vol *volume.Volinfo) ([]*Node, error) {
+	// Cluster graphs need to be linear and cannot have branches
+	// All xlators at a level in a cluster graph should be the same
+	if len(t.Children) > 1 {
+		return nil, ErrInvalidClusterGraphTemplate
+	}
+
+	var (
+		siblings, descendents []*Node
+		err                   error
+	)
+
+	if len(t.Children) == 1 {
+		descendents, err = processClusterGraph(t.Children[0], vol)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Special case for protocol/client
+	if t.Voltype == "protocol/client" {
+		return newClientNodes(vol)
+	}
+
+	sc := getChildCount(t.Voltype, vol)
+	j, k := 0, 0
+	var n *Node
+	for _, d := range descendents {
+		if j%sc == 0 {
+			n = NewNode()
+			n.Voltype = t.Voltype
+			n.Id = fmt.Sprintf("%s-%s-%d", vol.Name, t.Id, k)
+			if err := setOptions(n, vol); err != nil {
+				_ = err
```

Comment:
what about this?

also, are there tests?

Reply:
I need to handle this as well. Thanks for reminding.

We've begun making [changes to xlators](https://review.gluster.org/#/q/project:glusterfs+branch:experimental) that will help us get started with the integration with xlators. I mainly pushed out this patch so that we can at least test the changes that are being made to xlators can be seen. It is just expected to dump out a volfile now, and there is no expectation that the volfile works (and it doesn't work). It mainly fails because xlators have certain options that are expected to be set by glusterd, but those options are not a part of the volume set framework (in both GD1 and GD2). They had custom code within GD1's volgen to set them. This PR will help us identify those options, which will help us create a standard mechanism to set such options.

No tests yet because the only way available to validate a generated volfile is to have glusterfsd load it. The generated volfiles aren't loadable, so any test would always fail.